src/msw/thread.cpp

```diff
@@ -1121,6 +1122,80 @@ wxThreadError wxThread::Kill()
     return rc;
 }
 
+// This function requires at least MSVC 2017 version 15.6 and
+// Windows 10 version 1607 to succeed.
+static void wxSetThreadNameOnWindows10(const wxChar *threadName)
+{
+    typedef HRESULT(WINAPI* SetThreadDescription_t)(HANDLE, PCWSTR);
+    static SetThreadDescription_t s_pfnSetThreadDescription = NULL;
+
+    static bool s_initDone = false;
+    if ( !s_initDone )
+    {
+        wxLoadedDLL dllKernel32("kernel32.dll");
+        wxDL_INIT_FUNC(s_pfn, SetThreadDescription, dllKernel32);
+        s_initDone = true;
+    }
+
+    if ( s_pfnSetThreadDescription ) {
+        if ( FAILED((*s_pfnSetThreadDescription)(
+            GetCurrentThread(),
+            threadName)) )
+        {
+            wxLogSysError("SetThreadDescription failed");
+        }
+    }
+}
+
+// This function works with all MSVC versions.
+static void wxSetThreadNameOnAnyMSVC(const char* threadName)
+{
+    // This implementation is taken almost verbatim from:
+    // https://msdn.microsoft.com/en-us/library/xcb2z8hs.aspx
+
+    const DWORD MS_VC_EXCEPTION = 0x406D1388;
+#pragma pack(push,8)
+    typedef struct tagTHREADNAME_INFO
+    {
+        DWORD dwType; // Must be 0x1000.
+        LPCSTR szName; // Pointer to name (in user addr space).
+        DWORD dwThreadID; // Thread ID (-1=caller thread).
+        DWORD dwFlags; // Reserved for future use, must be zero.
+    } THREADNAME_INFO;
+#pragma pack(pop)
+
+    THREADNAME_INFO info;
+    info.dwType = 0x1000;
+    info.szName = threadName;
+    info.dwThreadID = (DWORD)-1;
+    info.dwFlags = 0;
+#pragma warning(push)
+#pragma warning(disable: 6320 6322)
+    __try {
+        RaiseException(MS_VC_EXCEPTION, 0,
+            sizeof(info) / sizeof(ULONG_PTR), (ULONG_PTR*)&info);
+    }
+    __except (EXCEPTION_EXECUTE_HANDLER) {
+    }
+#pragma warning(pop)
+}
+
+void wxThread::SetName(const wxString &name)
+{
+    wxCHECK_RET(this == This(),
+        "SetName() must be called in the context of the thread to be named");
+
+    // because the name argument is most often expected to to be
+    // something like "MyThread::Entry", drop everything after
+    // and including the first ':' to get the bare class name
+    const wxString trimmedName(name.BeforeFirst(':'));
```

Comment:
> I think this is surprising and rather counter-intuitive,

The reason it works that way is that I intended `SetName` to be nearly always used with `__PRETTY_FUNCTION__` or similar as argument.

> I want to be able to set my thread name to exactly the name given, which could be "Worker: 1" (if I also have "Worker: 2", ...). I.e. I don't think we should do anything at all with the provided name, but just use it as is.

Perhaps there could be two variants of `SetName` then, one for setting literally what the caller asked for, and another one that does some trimming... For with the 15-character limit on Linux, the first 15 characters of `__PRETTY_FUNCTION__` result in all threads being named "virtual wxThrea", which is not very useful.

> This means we don't actually need `__PRETTY_FUNCTION__` etc neither.

We don't _need_ it per se, but I'm assuming in a lot of cases having the derived class name as the thread name is good enough for most people, and if it can be acquired without explicitly typing it as a string literal, all the better.

Reply:
> > I think this is surprising and rather counter-intuitive,
> 
> The reason it works that way is that I intended `SetName` to be nearly always used with `__PRETTY_FUNCTION__` or similar as argument.

I really think this should be done in your own code and not imposed on all uses of the library. I.e. I almost never want to use the function name, however pretty it may be, for my  thread names because I give them more distinctive names, such as "event-dispatcher" or "computation $N" or "downloader" etc.

> Perhaps there could be two variants of `SetName` then, one for setting literally what the caller asked for, and another one that does some trimming... For with the 15-character limit on Linux, the first 15 characters of `__PRETTY_FUNCTION__` result in all threads being named "virtual wxThrea", which is not very useful.

This seems like another excellent argument in favour of using the user-provided name directly without processing it in any way.
 
> > This means we don't actually need `__PRETTY_FUNCTION__` etc neither.
> 
> We don't _need_ it per se, but I'm assuming in a lot of cases having the derived class name as the thread name is good enough for most people, and if it can be acquired without explicitly typing it as a string literal, all the better.

I'm sorry, I really don't see how can
```cpp
SetName(wxTHREAD_NAME);
```
be better than
```cpp
SetName("my name");
```
and IMO it's significantly worse. Of course, absolutely nothing prevents you from doing this in your code, but it really shouldn't be the recommended way to use `SetName()`.